core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -1798,8 +1791,16 @@ public Optional<GetExchangeSinkInstanceHandleResult> getExchangeSinkInstanceHand
                     attempt));
         }
 
-        public Optional<RemoteTask> schedule(int partitionId, ExchangeSinkInstanceHandle exchangeSinkInstanceHandle, int attempt, InternalNode node, boolean speculative)
+        public Optional<RemoteTask> schedule(int partitionId, ExchangeSinkInstanceHandle exchangeSinkInstanceHandle, int attempt, NodeLease nodeLease, boolean speculative)
         {
+            InternalNode node;
+            try {
+                node = Futures.getDone(nodeLease.getNode());
+            }
+            catch (ExecutionException e) {
+                throw new UncheckedExecutionException(e);
```

Comment:
Maybe add a message to make it obvious that the `nodeLease` is expected to be "done" here